File: integration-test/test/MetadataSpec.js
'use strict'

const asUserWithDocumentSet = require('../support/asUserWithDocumentSet')
const ApiBrowser = require('../lib/ApiBrowser')

function findInput(fieldName, wait) {
  const ret = { css: `tr[data-field-name="${fieldName}"] input` }
  if (wait) ret.wait = wait
  return ret
}

describe('Metadata', function() {
  asUserWithDocumentSet('Metadata/basic.csv', function() {
    beforeEach(async function() {
      this.browser.loadShortcuts('api')
    })

    describe('the Show page', function() {
      beforeEach(async function() {
        this.locator = findInput('foo')
        this.locatorWithWait = findInput('foo', 'fast')

        await this.browser.click({ tag: 'h3', contains: 'First' })
        await this.browser.sleep(500) // animation -- the "tag" button moves
        await this.browser.click({ link: 'Fields', wait: 'fast' })
        await this.browser.find(this.locatorWithWait)
      })

      it('should show default metadata', async function() {
        const value = await this.browser.getAttribute(this.locator, 'value')
        expect(value).to.eq('foo0')
      })

      it('should show new metadata when browsing to a new document', async function() {
        await this.browser.click('.document-nav a.next')
        const value = await this.browser.getAttribute(this.locatorWithWait, 'value')
        expect(value).to.eq('foo1')
      })

      it('should modify metadata', async function() {
        await this.browser.clear(this.locator)
        await this.browser.sendKeys('newFoo', this.locator)
        await this.browser.click('.document-nav a.next')
        await this.browser.find(this.locatorWithWait)
        await this.browser.click('.document-nav a.previous')

        const value = await this.browser.getAttribute(this.locatorWithWait, 'value')
        expect(value).to.eq('newFoo')
      })

      it('should add and remove metadata fields', async function() {
        await this.browser.click({ link: 'Organize fields' })
        await this.browser.click({ button: 'Add Field', wait: true })
        await this.browser.sendKeys('baz', '.metadata-schema-editor tbody tr:last-child input')
        await this.browser.click({ button: 'Close' })
        await this.browser.sendKeys('a baz value', findInput('baz'))

        // Navigate away and come back
        await this.browser.click('.document-nav a.next')
        await this.browser.click('.document-nav a.previous')
        const value = await this.browser.getAttribute(findInput('baz', 'fast'), 'value')
        expect(value).to.eq('a baz value')

        // Remove the field
        await this.browser.click({ link: 'Organize fields' })
        await this.browser.click('.metadata-schema-editor tbody tr:last-child a.remove')
        await this.browser.alert().accept()
        await this.browser.click({ button: 'Close' })

        // Navigate away and come back, to ensure request was sent
        await this.browser.click('.document-nav a.next')
        await this.browser.click('.document-nav a.previous')
        await this.browser.assertExists(this.locator, { wait: true })
      })

      it('should search for metadata', async function() {
        await this.browser.sendKeys("foo:foo0 OR bar:bar1", '.search input[name=query]')
        await this.browser.click('.search button')

        const text = await this.browser.getText({ css: '#document-list ul.documents', wait: 'pageLoad' })
        expect(text).to.match(/First/)
        expect(text).to.match(/Second/)
        expect(text).not.to.match(/Third/)
      })

      it('should warn when searching invalid metadata columns', async function() {
        await this.browser.sendKeys("foo2:x", '.search input[name=query]')
        await this.browser.click('.search button')

        const text = await this.browser.getText({ css: '#document-list ul.warnings:not(:empty)', wait: 'pageLoad' })
        expect(text).to.match(/There is no “foo2” field/)
        expect(text).to.match(/text/) // a hard-coded field
        expect(text).to.match(/bar/) // a valid field
      })
    })

    describe('GET /documents', function() {
      it('should return metadata when requested', async function() {
        const url = await this.browser.getUrl()
        const m = /^([^:]+:\/\/[^\/]+)\/documentsets\/(\d+)/.exec(url)
        if (!m) throw new Error(`We're at the wrong URL: ${url}`)

        const host = m[1]
        const id = m[2]

        const apiToken = await this.browser.shortcuts.api.createApiToken(id, 'metadata')

        const api = new ApiBrowser({ baseUrl: `${host}/api/v1`, apiToken: apiToken })
        const body = await api.GET(`/document-sets/${id}/documents?fields=metadata`)
        // Body items are sorted by title, and titles are alphabetical in basic.csv
        expect(body.items.map(i => i.metadata)).to.deep.eq([
          { foo: 'foo0', bar: 'bar0' },
          { foo: 'foo1', bar: 'bar1' },
          { foo: 'foo2', bar: 'bar2' },
        ])
      })
    })
  })
})

File: integration-test/spec/metadata_spec.rb
#!/usr/bin/env ruby

require './spec/spec_helper'

describe 'Metadata' do
  before do
    @user = admin_session.create_test_user
    page.log_in_as(@user)
    page.create_document_set_from_csv('files/metadata-spec/basic.csv')
  end

  after do
    admin_session.destroy_test_user(@user)
  end

  describe 'documentsets/:id/show' do
    before do
      page.open_document_in_list_with_name('First')
      page.waiting_for_css_transitions_on_selector('.document-metadata') do
        page.click_link('Fields')
      end
    end

    it 'should show metadata' do
      assert page.has_field?('foo', with: 'foo0', wait: WAIT_FAST)
    end

    it 'should change metadata when browsing to the next document' do
      page.within('#document-current') do
        page.click_link('Next')
      end
      assert page.has_field?('foo', with: 'foo1', wait: WAIT_FAST)
    end

    it 'should modify metadata' do
      page.within('.document-metadata') do
        page.fill_in('foo', with: 'newFoo')
      end

      # Navigate away and back
      page.within('#document-current') do
        page.click_link('Next')
        page.click_link('Previous')
      end

      assert page.has_field?('foo', with: 'newFoo', wait: WAIT_FAST)
    end

    it 'should add/remove metadata fields' do
      # Add a field
      page.click_link('Organize fields…')
      page.click_button('Add Field', wait: WAIT_FAST)
      page.within('.metadata-schema tbody tr:last-child') do
        page.fill_in('name', with: 'baz')
      end
      page.click_button('Close')
      # Wait for everything to close
      page.assert_no_selector('.metadata-schema', wait: WAIT_FAST)

      # Set the field value
      page.fill_in('baz', with: 'a baz value')

      # Navigate away and back
      page.within('#document-current') do
        page.click_link('Next')
        page.click_link('Previous')
      end

      page.assert_no_selector('.transaction-queue-communicating', wait: WAIT_FAST)
      assert_equal('a baz value', page.find(:fillable_field, 'baz').value)

      # Remove the field
      page.click_link('Organize fields…')
      page.within('.metadata-schema tbody', wait: WAIT_FAST) do # wait for dialog to load
        page.accept_confirm(wait: WAIT_FAST) do # wait for modal to appear after click
          page.within('tr:last-child') do
            page.click_link('Delete')
          end
        end
      end
      page.click_button('Close')
      # Wait for everything to close
      page.assert_no_selector('.metadata-schema', wait: WAIT_FAST)

      # Navigate away and back
      page.within('#document-current') do
        page.click_link('Next')
        page.click_link('Previous')
      end

      page.assert_no_selector('.transaction-queue-communicating', wait: WAIT_FAST)
      # Assert the field is gone
      assert !page.has_field?('baz')
    end

    it 'should search by metadata' do
      page.search_for_q('foo:foo0 OR bar:bar1')

      page.assert_selector('#document-list li', text: 'First')
      page.assert_selector('#document-list li', text: 'Second')
      page.assert_no_selector('#document-list li', text: 'Third')
    end

    it 'should warn on invalid metadata column search' do
      page.search_for_q('foo2:x')

      page.within('#document-list') do
        page.assert_selector('ul.warnings', text: /There is no “foo2” field/)
        page.assert_selector('ul.validFieldNames li', text: 'text') # suggests a built-in field type
        page.assert_selector('ul.validFieldNames li', text: 'bar')  # suggests an existing metadata field
      end
    end
  end

  it 'should return metadata from the API' do
    api = page.create_api_browser
    document_set_id = page.current_path.split(/\//)[2]
    res = api.GET("/document-sets/#{document_set_id}/documents?fields=metadata")

    # Body items are sorted by title, and titles are alphabetical in basic.csv
    assert_equal([
      { 'foo' => 'foo0', 'bar' => 'bar0' },
      { 'foo' => 'foo1', 'bar' => 'bar1' },
      { 'foo' => 'foo2', 'bar' => 'bar2' },
    ], JSON.load(res.body)['items'].map{ |item| item['metadata'] })
  end
end
